internal/fs/hns_bucket_test.go

```diff
@@ -100,3 +100,101 @@ func (t *HNSBucketTests) TestDeleteFolder() {
+	newDirPath := path.Join(mntDir, "foo_rename")
+
+	err = os.Rename(oldDirPath, newDirPath)
+
+	assert.NotNil(t.T(), err)
+	_, err = os.Stat(newDirPath)
+	assert.NotNil(t.T(), err)
+}
+
+func (t *HNSBucketTests) TestRenameFolderWithDstDirectoryIsNotEmpty() {
+	oldDirPath := path.Join(mntDir, "foo")
+	_, err = os.Stat(oldDirPath)
+	assert.NoError(t.T(), err)
+	newDirPath := path.Join(mntDir, "bar")
+	_, err = os.Stat(newDirPath)
+	assert.NoError(t.T(), err)
+
+	err = os.Rename(oldDirPath, newDirPath)
+
+	assert.NotNil(t.T(), err)
+}
+
+func (t *HNSBucketTests) TestRenameFolderWithSameParent() {
+	oldDirPath := path.Join(mntDir, "foo")
+	_, err = os.Stat(oldDirPath)
+	assert.NoError(t.T(), err)
+	newDirPath := path.Join(mntDir, "foo_rename")
+
+	err = os.Rename(oldDirPath, newDirPath)
+
+	assert.NoError(t.T(), err)
+	_, err = os.Stat(oldDirPath)
+	assert.NotNil(t.T(), err)
+	_, err = os.Stat(newDirPath)
+	assert.NoError(t.T(), err)
+	dirEntries, err := os.ReadDir(newDirPath)
+	assert.NoError(t.T(), err)
+	assert.Equal(t.T(), 5, len(dirEntries))
+	for i := 0; i < 5; i++ {
+		switch dirEntries[i].Name() {
+		case "test":
+			assert.Equal(t.T(), "test", dirEntries[i].Name())
+			assert.True(t.T(), dirEntries[i].IsDir())
+		case "test2":
+			assert.Equal(t.T(), "test2", dirEntries[i].Name())
+			assert.True(t.T(), dirEntries[i].IsDir())
+		case "file1.txt":
+			assert.Equal(t.T(), "file1.txt", dirEntries[i].Name())
+			assert.False(t.T(), dirEntries[i].IsDir())
+		case "file2.txt":
+			assert.Equal(t.T(), "file2.txt", dirEntries[i].Name())
+			assert.False(t.T(), dirEntries[i].IsDir())
+		case "implicit_dir":
+			assert.Equal(t.T(), "implicit_dir", dirEntries[i].Name())
+			assert.True(t.T(), dirEntries[i].IsDir())
+		}
+	}
+}
+
+func (t *HNSBucketTests) TestRenameFolderWithDifferentParent() {
+	oldDirPath := path.Join(mntDir, "foo")
+	_, err = os.Stat(oldDirPath)
+	assert.NoError(t.T(), err)
+	newDirPath := path.Join(mntDir, "bar", "foo_rename")
+
+	err = os.Rename(oldDirPath, newDirPath)
+
+	assert.NoError(t.T(), err)
+	_, err = os.Stat(oldDirPath)
+	assert.NotNil(t.T(), err)
+	_, err = os.Stat(newDirPath)
+	assert.NoError(t.T(), err)
+	dirEntries, err := os.ReadDir(newDirPath)
+	assert.NoError(t.T(), err)
+	assert.Equal(t.T(), 5, len(dirEntries))
+	for i := 0; i < len(dirEntries); i++ {
+		switch dirEntries[i].Name() {
+		case "test":
+			assert.Equal(t.T(), "test", dirEntries[i].Name())
+			assert.True(t.T(), dirEntries[i].IsDir())
+		case "test2":
+			assert.Equal(t.T(), "test2", dirEntries[i].Name())
+			assert.True(t.T(), dirEntries[i].IsDir())
+		case "file1.txt":
+			assert.Equal(t.T(), "file1.txt", dirEntries[i].Name())
+			assert.False(t.T(), dirEntries[i].IsDir())
+		case "file2.txt":
+			assert.Equal(t.T(), "file2.txt", dirEntries[i].Name())
+			assert.False(t.T(), dirEntries[i].IsDir())
+		case "implicit_dir":
+			assert.Equal(t.T(), "implicit_dir", dirEntries[i].Name())
+			assert.True(t.T(), dirEntries[i].IsDir())
+		}
+	}
+}
```

Comment:
We also need tests for the following scenarios I think (please correct me if we have those have covered already).

1. When source directory has files opened. Here, rename should fail.
2. When destination directory exists and is emoty. Here rename should pass.
3. When destination directory path is a parent/ancestor of the given source directory path. And the reverse. Both these cases, rename should fail.
4. When both source and destination directory are the same. Here rename should pass without doing anything.

Reply:
2. I thought to write, but `os.Rename` fails even if on empty dstDirectory. It works with `mv` command only
3 & 4 Is not even possible from file system, it will not even come to rename layer. Also 4 will be similar to 2 only sort of to as os.Rename does not support if dst directory even exist.

Reply:
For 1 I have added the test. Thanks.